### ISSUE TITLE
Debug stepping skips over assert() statements

### ISSUE THREAD

**Issue by mjross** [Komodo/KomodoEdit#3745]
### Summary
Debug stepping (both Debug > Step In and Debug > Step Over) jumps over all assert() statements, regardless whether the values they test are true or false, and even assert() statements that call other functions, thereby preventing me from stepping into those called functions. If I step into a function that consists of assert() statements only, then the Debugger jumps straight to the end of the function.

It is possible that this behavior now occurs with statements other than assert(). 

### Steps to Reproduce
Help > Troubleshooting > Restart in Safe-Mode. 

Run some example code:
```
function test() {
    assert( true );
    echo "1\n";
    assert( false );
    echo "2\n";
    assert( return_true() );
    echo "3\n";
}
function return_true() {
    return true;
}
test();
```

When I step into test(), the Debugger skips the first assert() and goes straight to the first echo statement. Then doing either Debug > Step In or Debug > Step Over, the Debugger skips the second assert() and goes straight to the second echo statement. It behaves the same on the third pair, so I can't even step into return_true().

### Expected Results
I naturally expected Komodo to behave as in the past, allowing me to step forward one statement at a time.

### Actual Results
Please see above.

### Platform Information
Komodo IDE 11.1.1 build 91089 (on Windows Home 10.0.17763.437)

### Attachments
![Clipboard01](https://user-images.githubusercontent.com/829492/56397923-eb19a700-620b-11e9-9090-64b712143f97.jpg)

[pystderr.log.txt](https://github.com/Komodo/KomodoEdit/files/3096528/pystderr.log.txt)

**Comment by th3coop** [comment]
I'm not able to reproduce this.  What version of PHP (I assume it's PHP) are you running?

**Comment by mjross** [comment]
"PHP 7.3.2 (cli) (built: Feb  5 2019 23:16:38) ( ZTS MSVC15 (Visual C++
2017) x86 )"

**Comment by th3coop** [comment]
K i'll get 7.3 and see if that changes anything.

**Comment by th3coop** [comment]
Oh we don't have 7.3 bits.  Did you stick xdebug original in there?

**Comment by mjross** [comment]
I'm not sure what you mean by 7.3 bits. The PHP I installed on my new laptop in February was PHP 7.3.2 VC15 32-bit Thread Safe. assert() statements were working fine until the day I submitted this issue. I don't recall making any changes to my PHP environment, but will be happy to answer any questions to help resolve this (for me, assert() statements are essential).

xdebug original? I'm not sure what you mean by that. When I installed PHP 7.3.2, I also installed php_xdebug-2.7.0RC2-7.3-vc15.dll into the extensions directory (in my case, C:\_a\PHP\ext) and reference it in my php.ini using the line 
`zend_extension = php_xdebug-2.7.0RC2-7.3-vc15.dll`
I also have the lines:
```
xdebug.remote_connect_back = true
xdebug.remote_enable = true
xdebug.remote_host = 127.0.0.1
xdebug.remote_handler = dbgp
xdebug.remote_mode = req
xdebug.remote_port = 9000
```
I vaguely recall I got those lines from a Reddit post helping people get Xdebug. If I don't include the above, then Komodo complains...

![Clipboard01](https://user-images.githubusercontent.com/829492/57264808-b5612480-7039-11e9-9343-74cfd6e3a48d.jpg)

![Clipboard02](https://user-images.githubusercontent.com/829492/57264811-b85c1500-7039-11e9-96f8-43a7b3a2e4af.jpg)

**Comment by th3coop** [comment]
@mjross, we ship xdebug.  That's what i mean by 7.3 bits.  The xdebug we ship doesn't support 7.3.  Original Xdebug ==  not the one we build and ship.
How are you including this Php.ini?  Is it in your PHP install directory with you php.exe?

**Comment by mjross** [comment]
I think the phrase "7.3 bits" was throwing me off because I immediately thought of 32- versus 64-bit versions of PHP.

To answer your questions, yes, my php.ini is in the root PHP installation directory (for me that is C:_a\PHP\ext), and presumably PHP automatically sees that file whenever it runs.

When you say that you ship Xdebug, does that mean that when I install Komodo IDE it would add Xdebug for me and I wouldn't need to do it manually? And should I be using an earlier version of PHP? When I got my new laptop recently, I simply downloaded and installed the most recent stable version.

**Comment by th3coop** [comment]
It's not added, it ships with Komodo.  When you point Komodo at a PHP interpreter, it will figure out some things about it to determine which Xdebug it should point at in the Komodo install directory.  If you use PHP 7.2 it will work out of the box without the custom stuff you did.

Having said that, you're not doing anything wrong.  Using any properly compiled Xdebug (ie for your PHP install) should work.  I created a ticket to build 7.3 binaries but there won't be a rush to get that out over other bugs.

How do you normally install PHP?  Do you use WAMP or XAMPP?  If yes, you probably already have 7.2 installed and can just change the interpreter Komodo is pointing at in your prefs.

**Comment by mjross** [comment]
Thanks for the clarification. I had no idea that I should have installed
PHP 7.2 instead of 7.3. To answer your question, I don't use any of those
packages, such as WAMP, and instead prefer installing each component on its
own. So I simply ran the usual PHP installer for Windows. It sounds like it
would be best for me to simply uninstall 7.3 and install 7.2. There's
nothing special about 7.3 that I need. I just chose what I thought was the
latest and greatest. How does that sound?

**Comment by th3coop** [comment]
> How does that sound?

as long as it resolves your problem!  Keep me posted.

**Comment by mjross** [comment]
I replaced PHP 7.3.2 with 7.2.18, started Komodo, and went to Preferences to point Komodo to the new php.exe. But I'm still getting the same error as shown above. The log is below.

[pystderr.log.txt](https://github.com/Komodo/KomodoEdit/files/3154446/pystderr.log.txt)

**Comment by th3coop** [comment]
There's nothing about PHP in those logs.  Are you sure they are the right ones?  Are you still defining a custom php.ini file?  Remove it if you are.

**Comment by mjross** [comment]
Yes, I was defining a custom php.ini file. (I was assuming one is always needed by PHP to get started. Never tried it without.) So I removed it just now and restarted Komodo, but am still getting the error above.

I'm really glad you asked about the Komodo logs, because I now realize that the doskey command I was using to access the logs was referencing a user account I'm no longer using. Sorry about that. Here are the correct logs...

[pystderr.log.txt](https://github.com/Komodo/KomodoEdit/files/3154919/pystderr.log.txt)

**Comment by mjross** [comment]
It appears Komodo is looking for C:\_a\Komodo_IDE_11\lib\support\php\debugging\7.2\ts-vc**15**-x86\php_xdebug.dll, but in that directory is ts-vc**14**-x86\php_xdebug.dll.

**Comment by mjross** [comment]
Is there a way we could induce Komodo to look for the ts-vc14-x86\php_xdebug.dll?

**Comment by th3coop** [comment]
Yeah, the vc15 means that it was compiled with Visualstudios compiler 15.  I'm not even sure how we could have gotten into this state unless PHP changed which compiler they use to compile PHP 7.* part way through the minor releases.  When we build Xdebug for release, we download the PHP version the dll will be for and do some magic to check which compiler built it then use that compiler to use Xdebug.  

There could be a bug in our build scripts which I'll have to look at.  For YOU though, we;re going to come full circle, you can download the appropriate version of Xdebug from https://xdebug.org/download.php then either use a custom php.ini file to point at it like you were before OR drop it into a folder in Komodo's support dir: **C:\_a\Komodo_IDE_11\lib\support\php\debugging\7.2\ts-vc15-x86\php_xdebug.dll** should work.  Make sure you download the 32bit Xdebug.

**Comment by th3coop** [comment]
Fyi, filed this bug: https://github.com/Komodo/KomodoEdit/issues/3754

**Comment by mjross** [comment]
Thanks! Interesting. My first thought is that if I use the Xdebug from https://xdebug.org/download.php then I'll again be able to do basic debugging, but Debug stepping will still skip over assert() statements? If that's the case, would going all the way back to PHP 7.1 resolve it? (I'll test that first scenario after lunch. "Grub first, then ethics." :-)

**Comment by th3coop** [comment]
> but Debug stepping will still skip over assert() statements?

I'm hoping no, but that's possible.  I don't understand the skip assertions problem yet.  Ideally we would confirm that the issue is either resolved or not when you're using OUR compiled bits so switching to 7.1 would be awesome if you don't mind the extra work.  Looking through the patches we apply to xdebug, I dont see anything about assert statements so there shouldn't be anything that makes this work with our Xdebug and not work with the original.  This makes me think there is something unique with your setup.

**Comment by mjross** [comment]
Success! I added php_xdebug-2.7.2-7.2-vc15.dll to C:\_a\PHP\ext and added `zend_extension = php_xdebug-2.7.2-7.2-vc15.dll` to my php.ini. Preferences approved it and I'm even able to step into assert() statements. My personal dev work was on hold without that ability, so this is very cool. Thanks, @th3coop!

The 7.3 stuff you kindly noted in other issues, so unless I'm mistaken, I can close this issue. If I missed anything, feel free to reopen it.

**Comment by th3coop** [comment]
yep, all good to close this one!

**Comment by mjross** [comment]
As of now, PHP version 8.0.3 has been released, as well as 7.4.16 as the latest in the 7 series. Also, I upgraded my Komodo IDE to version 12.0.1. So I'm wondering what is the most recent version in the 7 series that I could upgrade my laptop to without breaking assert() statements in Komodo? In other words, in reference to the comment you added on 7 May 2019, have the binaries now been built for PHP 7.2, or 7.3, or even 7.4?
